designs/generic-git-provider-gitops.md

```diff
@@ -0,0 +1,205 @@
+    - Provider specific methods will be used for specific providers like GitHub, using the provider API to check the status of the remote.
+- if the repository does not exist, we determine if we can create it for the user; if we can, we will do so, and initialize it. 
+- if the repository does not exist, we will proceed as we do today, cloning and initializing it as needed.
+
+Preflight validations will be executed to check the existence of the repository at CLI run time; if it does not exist, and the provider does not support repository creation, we will return a validation error. We will perform another validation durning the flux installation tasks to take care of cases such as the repository being deleted between preflight validations and the start of the task.
+
+The CLI code will check for the presence of a git provider in the method [`BootstrapToolkitsComponents(ctx context.Context, cluster *types.Cluster, gitOpsConfig *v1alpha1.GitOpsConfig)`](https://github.com/aws/eks-anywhere/blob/main/pkg/executables/flux.go#L32), and execute a provider-specific sub method (e.g. `boostrapToolkitsComponentsGit`, `bootstrapToolkitsComponentsGithub`) depending on which provider field is present in the GitOpsConfig (e.g. `Github` is non-nil; `Git` is non-nil, etc). This will execute the logic needed to check for the existence of the repository, call the appropriate `flux bootstrap` command for our provider, and proceed to the generic cloning and initialization tasks.
+
+### `Git` GitOpsConfig  Spec
+I propose modifying the `GitOpsConfig`, to move the `fluxSystemNamespace` and `clusterConfigPath` fields up a level to be a part of `flux`, instead of reproducing them in both `git` and `github`. provider blocks. This will make for a more consistent experience across provider, all of which will require a `fluxSystemNamespace` and `clusterConfigPath`.
+```yaml
+apiVersion: anywhere.eks.amazonaws.com/v1alpha1
+kind: GitOpsConfig
+metadata:
+  name: my-generic-git-provider-gitops
+spec:
+  flux:
+    fluxSystemNamespace: ""
+    clusterConfigPath: ""
+    git:
+      repositoryUrl: myClusterGitopsRepo
+      username: myGitProviderUserName
+      passwordEnvironmentVariable: MY_PASSWORD_ENV_VAR
+      privateKeyFile: '/path/to/my/private_key.rsa'
+      branch: branchToUseInRemoteRepo
+```
+
+### Git Configuration Structs
+The `Flux` configuration will be modified to contain generic flux options `FluxSystemNamespace`, `Branch`, and `ClusterConfigPath`, which will be universal across git providers used in conjunction with Flux. A new field, `Git`, will be added, and validations will ensure that `Git` and `Github` (and any other Provider type) are mutually exclusive.
+```Go
+type Flux struct {
+    // Github is the name of the Git Provider to host the Git repo.
+	Github Github `json:"github,omitempty"`
+
+    // Git is the name of the generic Git Provider to host the Git repository
+    Git Git `json:"git,omitempty"`
```

Comment:
what do you think about having Git field at the level it is, but then Github as a sub field, and then another sub field to accept all fields required for a generic git provider. so as we go forward and add more specific implementations we can add them as sub fields under git.

Reply:
let's discuss at office hours today -- I've updated the API design to use a ref to a git provider specific configuration to make this more modular